test/e2e/storage/testsuites/capacity.go

```diff
@@ -0,0 +1,380 @@
+	// found during the preceding Match call.
+	MatchedCapacities() []storagev1beta1.CSIStorageCapacity
+}
+
+// HaveCapacitiesForClass filters all storage capacity objects in a *storagev1beta1.CSIStorageCapacityList
+// by storage class. Success is when when there is at least one.
+func HaveCapacitiesForClass(scName string) CapacityMatcher {
+	return &haveCSIStorageCapacities{scName: scName}
+}
+
+type haveCSIStorageCapacities struct {
+	scName             string
+	matchingCapacities []storagev1beta1.CSIStorageCapacity
+}
+
+var _ CapacityMatcher = &haveCSIStorageCapacities{}
+
+func (h *haveCSIStorageCapacities) Match(actual interface{}) (success bool, err error) {
+	capacities, ok := actual.(*storagev1beta1.CSIStorageCapacityList)
+	if !ok {
+		return false, fmt.Errorf("expected *storagev1beta1.CSIStorageCapacityList, got: %T", actual)
+	}
+	h.matchingCapacities = nil
+	for _, capacity := range capacities.Items {
+		if capacity.StorageClassName == h.scName {
+			h.matchingCapacities = append(h.matchingCapacities, capacity)
+		}
+	}
+	return len(h.matchingCapacities) > 0, nil
+}
+
+func (h *haveCSIStorageCapacities) MatchedCapacities() []storagev1beta1.CSIStorageCapacity {
+	return h.matchingCapacities
+}
+
+func (h *haveCSIStorageCapacities) FailureMessage(actual interface{}) (message string) {
+	return fmt.Sprintf("no CSIStorageCapacity objects for storage class %q", h.scName)
+}
+
+func (h *haveCSIStorageCapacities) NegatedFailureMessage(actual interface{}) (message string) {
+	return fmt.Sprintf("CSIStorageCapacity objects for storage class %q:\n%s",
+		h.scName,
+		strings.Join(formatCapacities(h.matchingCapacities), "\n"),
+	)
+}
+
+// HaveCapacitiesForClassAndNodes matches objects by storage class name. It finds
+// all nodes on which the driver runs and expects one object per node.
+func HaveCapacitiesForClassAndNodes(client kubernetes.Interface, driverName, scName, topologyKey string) CapacityMatcher {
+	return &haveLocalStorageCapacities{
+		client:      client,
+		driverName:  driverName,
+		match:       HaveCapacitiesForClass(scName),
+		topologyKey: topologyKey,
+	}
+}
+
+type haveLocalStorageCapacities struct {
+	client      kubernetes.Interface
+	driverName  string
+	match       CapacityMatcher
+	topologyKey string
+
+	matchSuccess          bool
+	topologyValues        []string
+	expectedCapacities    []storagev1beta1.CSIStorageCapacity
+	unexpectedCapacities  []storagev1beta1.CSIStorageCapacity
+	missingTopologyValues []string
+}
+
+var _ CapacityMatcher = &haveLocalStorageCapacities{}
+
+func (h *haveLocalStorageCapacities) Match(actual interface{}) (success bool, err error) {
+	h.topologyValues = nil
+	h.expectedCapacities = nil
+	h.unexpectedCapacities = nil
+	h.missingTopologyValues = nil
+
+	// First check with underlying matcher.
+	success, err = h.match.Match(actual)
+	h.matchSuccess = success
+	if !success || err != nil {
+		return
+	}
+
+	// Find all nodes on which the driver runs.
```

Comment:
This looks similar to the actual algorithm for computing `CSIStorageCapacity`. Are we running into the risk of reimplementing the logic we are testing here?

Ideally we should compute the number of expected objects based on configured expectations but that might be hard (for example to configure the topology segment distribution over nodes) and the existing test framework may not give enough information today to do this computation.

So if it's indeed close enough to a reimplementation, which would be hard to maintain, does it make sense to require that all LocalStorage drivers are expected to be installed on every node, and just use the node count to compute the expected number of Storage Capacity objects? The tradeoff here is some types of drivers will be excluded, but maybe this is necessary if (1) we don't want to reimplement the logic we are testing and (2) we can't configure expectations more easily.

Reply:
> This looks similar to the actual algorithm for computing `CSIStorageCapacity`. Are we running into the risk of reimplementing the logic we are testing here?

Saw that after writing my own comment in https://github.com/kubernetes/kubernetes/pull/100537#discussion_r600891985 - yes, that's exactly the risk. The "let's do a special case for local storage" approach came from the desire to keep the test code simple and easy to review. I also expect that most drivers which use that will be for local storage, so the test would already be useful with that limitation.

> Ideally we should compute the number of expected objects based on configured expectations but that might be hard (for example to configure the topology segment distribution over nodes) and the existing test framework may not give enough information today to do this computation.

Correct. The driver deployment itself would have to compute and configure this, which would be complex.

> So if it's indeed close enough to a reimplementation, which would be hard to maintain, does it make sense to require that all LocalStorage drivers are expected to be installed on every node, and just use the node count to compute the expected number of Storage Capacity objects?

I find that too limiting. PMEM-CSI cannot necessarily run on all nodes. I also don't find checking of the CSINode objects that hard or complex.